src/core/lang_source.ml

```diff
@@ -282,15 +331,44 @@ let add_operator =
       :: List.stable_sort compare proto
     in
     let f env =
+      let return_t, proto =
+        let generalized, t =
+          Typing.generalize ~level:(-1)
+            (Type.make
+               (Type.Tuple (return_t :: List.map (fun (_, t, _, _) -> t) proto)))
+        in
+        let t = Typing.instantiate ~level:(-1) ~generalized t in
+        match t.Type.descr with
+          | Type.Tuple (return_t :: proto_t) ->
+              ( return_t,
+                List.map2 (fun (name, _, _, _) typ -> (name, typ)) proto proto_t
+              )
+          | _ -> assert false
+      in
+      (* Negotiate content for all sources and formats in the arguments. *)
+      let () =
+        let positions = Hashtbl.create 10 in
+        let position name =
+          let pos = try Hashtbl.find positions name with Not_found -> 1 in
+          Hashtbl.replace positions name (pos + 1);
+          pos
+        in
+        List.iter
+          (fun (name, typ) ->
+            let v =
+              try Some (assoc name (position name) env) with Not_found -> None
+            in
+            match v with None -> () | Some v -> check_content v typ)
+          proto
```

Comment:
Wouldn't it be easier to look for the nth argument in the environment (where n is the position in the proto, without taking the label in account) rather than bookkeeping this hastable for arguments with same label? I.e. use `List.iteri` instead of `List.iter`.

Reply:
Yeah I overthought this.